Add Inline landed costing feature

HKuz opened the pull request:
Addresses #4 

Opening PR to track progress - for now I see this as two 'phases' - 1) get the feature working on a standalone basis in v14 and 2) integrate the feature to work with WO subcontracting.

PHASE I
- [x] Update Inventory Tools Settings, PR, and PI schemas to include new necessary fields
- Purchase Receipt:
  - [x] Add setting-based server-side functionality to custom Inv tools Purchase Receipt
  - [x] Port JS/UI functionality to distribute landed costs and show/hide relevant columns
- Purchase Invoice:
  - [x] Add setting-based server-side functionality to custom Inv tools Purchase Invoice
  - [x] Port JS/UI functionality to distribute landed costs and show/hide relevant columns
- Testing (done manually for now) to see if everything working as expected:
  - [x] ERPNext current functionality still works: Turn enable inline LC feature OFF, submit PR, create LCV
  - [x] ERPNext current functionality still works: Turn enable inline LC feature ON, submit PR, create LCV
  - [x] ERPNext current functionality still works: Cancel LCV in prior test
  - [x] Feature: With feature ON, submit PR with inline LCs
  - [x] Feature: With feature ON, submit PI with update stock checked and with inline LCs
  - [x] Feature: With feature ON, submit a PR, create a PI off it (so can't check update stock), add inline LCs
  - [x] Feature: Cancel the PI in prior test
  - [ ] Feature: With feature ON, submit PR with inline LCs, create a PI off it and add additional LCs
- [ ] Documentation

PHASE II and tests -> moved to a separate issue

HKuz committed:
feat: add landed costing to settings

HKuz committed:
feat: add inline LC to purchase receipt UI

HKuz committed:
feat: add purchase invoice changes

HKuz committed:
feat: add landed_costing valuation rate override

HKuz committed:
feat: add inline LC feat to PR, PI and tie to settings

agritheory commented:
@HKuz I tried to rebase this but I don't feel confident in the decisions I was making, can I get you to rebase this and we'll split the remaining work into separate ticket?

HKuz committed:
Merge branch 'version-14' into inline_lc

HKuz commented:
@agritheory I rebased this, but in testing I realized at least one function needs to be updated for version-14 changes (it's trying to access a default account that isn't listed in Company settings anymore). I can ping you when this is updated and ready.

agritheory commented:
@HKuz Just had a look at this again, we need to add docs but should probably split out the tests into another ticket.

HKuz committed:
feat: update for erpnext code changes

HKuz committed:
Merge branch 'version-14' into inline_lc

HKuz commented:
All code is now updated and re-tested for changes in ERPNext's underlying functions. In running through the test scenarios above, I added another one to create a PR with LCs, then create a PI off it with additional LCs. This uncovered a bug where it's double-counting the LCs when the PI is submitted. 

In our JS code, when the Distribute Landed Costs Based On field is changed to either Qty or Amount, it's assuming *everything* in the taxes and charges table should be allocated (automatically sets the category to Valuation and Total and includes in the sum for the items table). It wasn't allowing the user to exclude a row (like sales tax in a PI) and recalculate the landed cost amount.

UPDATE - adjustments made in the javascript to let the user change the tax category and it triggers a re-calc of the landed cost amounts (excluding anything not marked as Valuation or Valuation and Total). Docs are also updated to note this.

For the double-counting solution I need to add a check in the PI for any LCs already in the PR so they're not double-counted.

HKuz committed:
docs: start inline landed costing

HKuz committed:
feat: allow change of tax category
